### PR TITLE
 Go To Statement Considered Harmful

### PR DESCRIPTION
**Pull request by shyouhei**
This pull request is a set of changes that reduce `goto`s.

**Note** that it doesn't kill everything.  For instance I see `goto`s written in `dln.c` are very legit; they are used to escape from functions, with cleaning allocated resources up.  I don't see any problem at all.  I didn't touch use cases like them.

However there are `goto`s that are questionable.  For instance

https://github.com/ruby/ruby/blob/eed4aaa0bf8aa3109936c4276a71c0cdab4d54d6/process.c#L1935-L1956

Why is this `goto` needed?  `0` should go to the `else` branch with or without this `goto`.  This is just complicating the source code for no reason.  Should just avoid.

### CODE REVIEW

File: class.c [hunk 1]
```diff
@@ -1782,26 +1778,25 @@ singleton_class_of(VALUE obj)
 {
     VALUE klass;
 
-    if (FIXNUM_P(obj) || FLONUM_P(obj) || STATIC_SYM_P(obj)) {
-      no_singleton:
+    switch (TYPE(obj)) {
```
Comment (nobu) [inline class.c:1781]: I think the previous code generated better code than `TYPE()`.

Comment (shyouhei) [reply class.c:1781]: @nobu tell us how better?  I don't see any clear problem

```diff
--- miniruby.master 2020-06-22 14:11:30.015038077 +0900
+++ miniruby        2020-06-22 14:11:18.846865697 +0900
@@ -1,102 +1,116 @@
 
-miniruby.master:	file format ELF64-x86-64
+miniruby:	file format ELF64-x86-64
 
 
 Disassembly of section .text:
 
 singleton_class_of:
+ 55                                    	pushq	%rbp
  41 57                                 	pushq	%r15
  41 56                                 	pushq	%r14
  53                                    	pushq	%rbx
+ 50                                    	pushq	%rax
  49 89 fe                              	movq	%rdi, %r14
- 41 80 fe 0c                           	cmpb	$12, %r14b
- 0f 84 0c 01 00 00                     	je	268 <singleton_class_of+0x11e>
- 44 89 f0                              	movl	%r14d, %eax
- 83 e0 01                              	andl	$1, %eax
- 48 85 c0                              	testq	%rax, %rax
- 0f 85 fd 00 00 00                     	jne	253 <singleton_class_of+0x11e>
- 44 89 f0                              	movl	%r14d, %eax
- 83 e0 03                              	andl	$3, %eax
- 83 f8 02                              	cmpl	$2, %eax
- 0f 84 ee 00 00 00                     	je	238 <singleton_class_of+0x11e>
  41 f6 c6 07                           	testb	$7, %r14b
- 0f 85 a8 00 00 00                     	jne	168 <singleton_class_of+0xe2>
- 4c 89 f0                              	movq	%r14, %rax
- 48 83 e0 f7                           	andq	$-9, %rax
- 0f 84 9b 00 00 00                     	je	155 <singleton_class_of+0xe2>
+ 0f 95 c0                              	setne	%al
+ 48 f7 c7 f7 ff ff ff                  	testq	$-9, %rdi
+ 40 0f 94 c5                           	sete	%bpl
+ 40 08 c5                              	orb	%al, %bpl
+ 40 80 fd 01                           	cmpb	$1, %bpl
+ 75 21                                 	jne	33 <singleton_class_of+0x46>
+ 49 83 fe 13                           	cmpq	$19, %r14
+ 7f 3e                                 	jg	62 <singleton_class_of+0x69>
+ 4d 85 f6                              	testq	%r14, %r14
+ 75 66                                 	jne	102 <singleton_class_of+0x96>
+ 48 8d 05 19 1c 54 00                  	leaq	5512217(%rip), %rax
+ 48 8b 18                              	movq	(%rax), %rbx
+ 48 83 fb 08                           	cmpq	$8, %rbx
+ 0f 85 04 01 00 00                     	jne	260 <singleton_class_of+0x148>
+ eb 6a                                 	jmp	106 <singleton_class_of+0xb0>
  49 8b 06                              	movq	(%r14), %rax
  89 c1                                 	movl	%eax, %ecx
  83 e1 1f                              	andl	$31, %ecx
- 83 c1 fc                              	addl	$-4, %ecx
- 83 f9 10                              	cmpl	$16, %ecx
- 77 1b                                 	ja	27 <singleton_class_of+0x72>
- 48 8d 15 02 86 24 00                  	leaq	2393602(%rip), %rdx
- 48 63 0c 8a                           	movslq	(%rdx,%rcx,4), %rcx
- 48 01 d1                              	addq	%rdx, %rcx
- ff e1                                 	jmpq	*%rcx
+ 83 f9 15                              	cmpl	$21, %ecx
+ 0f 87 83 00 00 00                     	ja	131 <singleton_class_of+0xda>
+ ba 00 00 0e 00                        	movl	$917504, %edx
+ 0f a3 ca                              	btl	%ecx, %edx
+ 73 24                                 	jae	36 <singleton_class_of+0x85>
+ 49 83 fe 14                           	cmpq	$20, %r14
+ 74 08                                 	je	8 <singleton_class_of+0x6f>
+ eb 47                                 	jmp	71 <singleton_class_of+0xb0>
+ 49 83 fe 14                           	cmpq	$20, %r14
+ 75 52                                 	jne	82 <singleton_class_of+0xc1>
+ 48 8d 05 ca 1b 54 00                  	leaq	5512138(%rip), %rax
+ 48 8b 18                              	movq	(%rax), %rbx
+ 48 83 fb 08                           	cmpq	$8, %rbx
+ 0f 85 c5 00 00 00                     	jne	197 <singleton_class_of+0x148>
+ eb 2b                                 	jmp	43 <singleton_class_of+0xb0>
+ 83 f9 05                              	cmpl	$5, %ecx
+ 75 46                                 	jne	70 <singleton_class_of+0xd0>
  a9 00 00 00 20                        	testl	$536870912, %eax
- 0f 85 ac 00 00 00                     	jne	172 <singleton_class_of+0x11e>
+ 74 49                                 	je	73 <singleton_class_of+0xda>
+ e9 c0 00 00 00                        	jmp	192 <singleton_class_of+0x156>
+ 49 83 fe 08                           	cmpq	$8, %r14
+ 75 2b                                 	jne	43 <singleton_class_of+0xc7>
+ 48 8d 05 85 1b 54 00                  	leaq	5512069(%rip), %rax
+ 48 8b 18                              	movq	(%rax), %rbx
+ 48 83 fb 08                           	cmpq	$8, %rbx
+ 0f 85 98 00 00 00                     	jne	152 <singleton_class_of+0x148>
+ 48 8d 3d e0 68 24 00                  	leaq	2386144(%rip), %rdi
+ 4c 89 f6                              	movq	%r14, %rsi
+ 31 c0                                 	xorl	%eax, %eax
+ e8 24 47 06 00                        	callq	411428 <rb_bug>
+ 49 83 fe 34                           	cmpq	$52, %r14
+ 74 13                                 	je	19 <singleton_class_of+0xda>
+ 41 f6 c6 01                           	testb	$1, %r14b
+ e9 86 00 00 00                        	jmp	134 <singleton_class_of+0x156>
+ b8 10 04 30 00                        	movl	$3146768, %eax
+ 0f a3 c8                              	btl	%ecx, %eax
+ 72 7c                                 	jb	124 <singleton_class_of+0x156>
  49 8b 5e 08                           	movq	8(%r14), %rbx
  f6 c3 07                              	testb	$7, %bl
- 75 32                                 	jne	50 <singleton_class_of+0xad>
- 49 89 df                              	movq	%rbx, %r15
- 49 83 e7 f7                           	andq	$-9, %r15
- 74 29                                 	je	41 <singleton_class_of+0xad>
+ 75 32                                 	jne	50 <singleton_class_of+0x115>
+ 48 89 d8                              	movq	%rbx, %rax
+ 48 83 e0 f7                           	andq	$-9, %rax
+ 74 29                                 	je	41 <singleton_class_of+0x115>
  48 8b 03                              	movq	(%rbx), %rax
  89 c1                                 	movl	%eax, %ecx
  83 e1 1f                              	andl	$31, %ecx
  83 f9 1b                              	cmpl	$27, %ecx
- 74 1c                                 	je	28 <singleton_class_of+0xad>
+ 74 1c                                 	je	28 <singleton_class_of+0x115>
  25 00 10 00 00                        	andl	$4096, %eax
  48 85 c0                              	testq	%rax, %rax
- 74 12                                 	je	18 <singleton_class_of+0xad>
+ 74 12                                 	je	18 <singleton_class_of+0x115>
  be 01 0c 00 00                        	movl	$3073, %esi
  48 89 df                              	movq	%rbx, %rdi
- e8 98 e1 1f 00                        	callq	2089368 <rb_ivar_get>
+ e8 e0 c1 1f 00                        	callq	2081248 <rb_ivar_get>
  4c 39 f0                              	cmpq	%r14, %rax
- 74 1a                                 	je	26 <singleton_class_of+0xc7>
+ 74 13                                 	je	19 <singleton_class_of+0x128>
  4c 8b 7b 20                           	movq	32(%rbx), %r15
  4c 89 f7                              	movq	%r14, %rdi
- e8 27 dc ff ff                        	callq	-9177 <rb_make_metaclass>
+ e8 3f db ff ff                        	callq	-9409 <rb_make_metaclass>
  48 89 c3                              	movq	%rax, %rbx
  4c 89 78 20                           	movq	%r15, 32(%rax)
- 49 89 c7                              	movq	%rax, %r15
- 49 83 e7 f7                           	andq	$-9, %r15
  49 8b 06                              	movq	(%r14), %rax
+ 40 84 ed                              	testb	%bpl, %bpl
  25 00 08 00 00                        	andl	$2048, %eax
  f6 c3 07                              	testb	$7, %bl
- 75 03                                 	jne	3 <singleton_class_of+0xd7>
- 4d 85 ff                              	testq	%r15, %r15
+ 75 07                                 	jne	7 <singleton_class_of+0x13f>
+ 48 89 d9                              	movq	%rbx, %rcx
+ 48 83 e1 f7                           	andq	$-9, %rcx
  48 8b 0b                              	movq	(%rbx), %rcx
  48 09 c1                              	orq	%rax, %rcx
  48 89 0b                              	movq	%rcx, (%rbx)
- eb 33                                 	jmp	51 <singleton_class_of+0x115>
- 49 83 fe 14                           	cmpq	$20, %r14
- 74 1d                                 	je	29 <singleton_class_of+0x105>
- 49 83 fe 08                           	cmpq	$8, %r14
- 74 0e                                 	je	14 <singleton_class_of+0xfc>
- 4d 85 f6                              	testq	%r14, %r14
- 75 43                                 	jne	67 <singleton_class_of+0x136>
- 48 8d 05 be 3f 54 00                  	leaq	5521342(%rip), %rax
- eb 10                                 	jmp	16 <singleton_class_of+0x10c>
- 48 8d 05 8d 3f 54 00                  	leaq	5521293(%rip), %rax
- eb 07                                 	jmp	7 <singleton_class_of+0x10c>
- 48 8d 05 9c 3f 54 00                  	leaq	5521308(%rip), %rax
- 48 8b 18                              	movq	(%rax), %rbx
- 48 83 fb 08                           	cmpq	$8, %rbx
- 74 21                                 	je	33 <singleton_class_of+0x136>
  48 89 d8                              	movq	%rbx, %rax
+ 48 83 c4 08                           	addq	$8, %rsp
  5b                                    	popq	%rbx
  41 5e                                 	popq	%r14
  41 5f                                 	popq	%r15
+ 5d                                    	popq	%rbp
  c3                                    	retq
- 48 8d 05 5b 19 54 00                  	leaq	5511515(%rip), %rax
+ 48 8d 05 d3 f4 53 00                  	leaq	5502163(%rip), %rax
  48 8b 38                              	movq	(%rax), %rdi
- 48 8d 35 49 88 24 00                  	leaq	2394185(%rip), %rsi
- 31 c0                                 	xorl	%eax, %eax
- e8 9a 5b 06 00                        	callq	416666 <rb_raise>
- 48 8d 3d 52 88 24 00                  	leaq	2394194(%rip), %rdi
- 4c 89 f6                              	movq	%r14, %rsi
+ 48 8d 35 19 68 24 00                  	leaq	2385945(%rip), %rsi
  31 c0                                 	xorl	%eax, %eax
- e8 9e 66 06 00                        	callq	419486 <rb_bug>
- 66 0f 1f 84 00 00 00 00 00            	nopw	(%rax,%rax)
+ e8 62 3b 06 00                        	callq	408418 <rb_raise>
+ 66 90                                 	nop
```

Comment (shyouhei) [reply class.c:1781]: That was diff of `singleton_class_of`, both compiled using `clang-11 -O3`.

Comment (nobu) [reply class.c:1781]: I see, the change of `rb_type`, clang, or both, did the thing.

**Comment by shyouhei** [comment]
Do I need to wait for @nobu's another comment?

**Comment by headius** [comment]
I'd love to see [transcode_restartable0](https://github.com/ruby/ruby/blob/9e41a75255d15765648279629fd3134cae076398/transcode.c#L435) done without gotos, since duplicating that code in a language without gotos was [quite challenging](https://github.com/jruby/jcodings/blob/184ccc4ef1db5b03c560d45fb57c2cdcada5bfcb/src/org/jcodings/transcode/Transcoding.java#L122), but this is great to see happening anyway! 😄🎉

File: compile.c [hunk 2]
```diff
@@ -6660,12 +6660,7 @@ compile_redo(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node, in
 	    ADD_INSN(ret, line, putnil);
 	}
     }
-    else if (iseq->body->type == ISEQ_TYPE_EVAL) {
-      redo_in_eval:
-	COMPILE_ERROR(ERROR_ARGS "Can't escape from eval with redo");
-	return COMPILE_NG;
-    }
-    else if (ISEQ_COMPILE_DATA(iseq)->start_label) {
+    else if (iseq->body->type == ISEQ_TYPE_EVAL && ISEQ_COMPILE_DATA(iseq)->start_label) {
```
Comment (nobu) [inline compile.c:6663]: Are these expressions equivalent?
Seems I'm missing something...

**Comment by nobu** [comment]
I've read up to `ibf_dump_object_object` now.
Taking a breath.

**Comment by duerst** [comment]
> I'd love to see [transcode_restartable0](https://github.com/ruby/ruby/blob/9e41a75255d15765648279629fd3134cae076398/transcode.c#L435) done without gotos, since duplicating that code in a language without gotos was [quite challenging](https://github.com/jruby/jcodings/blob/184ccc4ef1db5b03c560d45fb57c2cdcada5bfcb/src/org/jcodings/transcode/Transcoding.java#L122), but this is great to see happening anyway! 😄🎉

I'd love to work on this one time. The original code had many less gotos, What it did was that if it realized that it didn't have enough input bytes to output a character, it would just pretend it never had been able to process any of the bytes of that character, and return in that state. I still think it should be possible to go back to such an implementation. But I don't know exactly when I'll have enough time to work on it.

File: dir.c [hunk 3]
```diff
@@ -2226,8 +2224,13 @@ glob_opendir(ruby_glob_entries_t *ent, DIR *dirp, int flags, rb_encoding *enc)
         }
         ruby_qsort(ent->sort.entries, ent->sort.count, sizeof(ent->sort.entries[0]),
                    glob_sort_cmp, NULL);
+        return ent;
+
+      nomem:
```
Comment (nobu) [inline dir.c:2229]: Just out of curiosity, why not the end of the function?

Comment (shyouhei) [reply dir.c:2229]: I thought this is the right place to put `nomem:` because all `goto nomem;` happens inside of this scope.

But the end of the function also seems perfect.  No strong opinion now.

**Comment by shyouhei** [comment]
Ooooh I overlooked transcode_restartable0.  Labels defined by macro expansions...

File: compile.c (hunk 2 above)
Comment (shyouhei) [reply compile.c:6663]: Ah, this is a mistake.  Thank you.  It should be:

```suggestion
    else if (iseq->body->type != ISEQ_TYPE_EVAL && ISEQ_COMPILE_DATA(iseq)->start_label) {
```

File: enum.c [hunk 4]
```diff
@@ -3848,122 +3848,130 @@ struct enum_sum_memo {
...
-sum_iter(VALUE i, struct enum_sum_memo *memo)
+sum_iter_normalize_memo(struct enum_sum_memo *memo)
 {
-    const int unused = (assert(memo != NULL), 0);
-
-    long n = memo->n;
-    VALUE v = memo->v;
-    VALUE r = memo->r;
+    assert(FIXABLE(memo->n));
+    memo->v = rb_fix_plus(LONG2FIX(memo->n), memo->v);
```
Comment (nobu) [inline enum.c:3854]: Is it intentional that `if (n != 0)` condition is removed?

File: error.c [hunk 5]
```diff
@@ -908,23 +908,23 @@ rb_check_typeddata(VALUE obj, const rb_data_type_t *data_type)
...
+        etype = RTYPEDDATA_TYPE(obj)->wrap_struct_name;
+    }
+    else {
+        return DATA_PTR(obj);
     }
-    return DATA_PTR(obj);
+
+    /* rb_obj_classname() cannot be used.  A class name can be non-ASCII. */
+    const char *expected = data_type->wrap_struct_name;
+    VALUE actual = (etype) ? rb_str_new_cstr(etype) : rb_obj_class(obj);
```
Comment (nobu) [inline error.c:925]: Creating a new string here?

File: eval.c [hunk 6]
```diff
@@ -804,44 +806,37 @@ static VALUE
...
       case 1:
 	exc = argv[0];
-	if (NIL_P(exc))
-	    break;
-	if (isstr) {
+        if (isstr &&! NIL_P(exc)) {
 	    mesg = rb_check_string_type(exc);
 	    if (!NIL_P(mesg)) {
 		mesg = rb_exc_new3(rb_eRuntimeError, mesg);
-		break;
```
Comment (nobu) [inline eval.c:821]: As `rb_exc_new3` should return an exception instance, isn't ok to return it here?

File: eval.c [hunk 7]
```diff
@@ -804,44 +806,37 @@ static VALUE
...
 	break;
       default:
         rb_error_arity(argc, 0, 3);
     }
-    if (argc > 0) {
-	if (!rb_obj_is_kind_of(mesg, rb_eException))
-	    rb_raise(rb_eTypeError, "exception object expected");
-	if (argc > 2)
-	    set_backtrace(mesg, argv[2]);
+    if (NIL_P(mesg)) {
```
Comment (nobu) [inline eval.c:829]: `NIL_P(mesg)`  is true only in the case of `isstr`, I think.

File: file.c [hunk 8]
```diff
@@ -5432,7 +5425,14 @@ rb_f_test(int argc, VALUE *argv, VALUE _)
...
-    goto unknown;
+  unknown:
+    /* unknown command */
+    if (ISPRINT(cmd)) {
+        rb_raise(rb_eArgError, "unknown command '%s%c'", cmd == '\'' || cmd == '\\' ? "\\" : "", cmd);
+    }
+    else {
+        rb_raise(rb_eArgError, "unknown command \"\\x%02X\"", cmd);
+    }
 }
```
Comment (nobu) [inline file.c:5436]: May be better to put `UNREACHABLE_RETURN` here.

File: numeric.c [hunk 9]
```diff
@@ -4066,6 +4071,25 @@ rb_int_positive_pow(long x, unsigned long y)
     return int_pow(x, y);
 }
 
+static VALUE
+fix_pow_inverted(VALUE x, VALUE minusb)
+{
+    if (x == INT2FIX(0)) {
+        rb_num_zerodiv();
```
Comment (nobu) [inline numeric.c:4078]: `UNREACHABLE_RETURN` needed?

File: object.c [hunk 10]
```diff
@@ -3374,33 +3378,35 @@ rb_convert_to_integer(VALUE val, int base, int raise_exception)
 {
     VALUE tmp;
 
+    if (base) {
+        tmp = rb_check_string_type(val);
```
Comment (nobu) [inline object.c:3382]: Strictly speaking, this changes the behavior when `to_str` is defined to `Float`, `Integer`, and `NilClass`, but probably it does not matter, I guess.

File: rational.c [hunk 11]
```diff
@@ -2439,24 +2440,27 @@ parse_rat(const char *s, const char *const e, int strict, int raise)
     if (nexp != ZERO) {
         if (INT_NEGATIVE_P(nexp)) {
             VALUE mul;
-            if (!FIXNUM_P(nexp)) {
```
Comment (nobu) [inline rational.c:2442]: How about `if (!FIXNUM_P(nexp) || RB_FLOAT_TYPE_P(mul = f_expt10(LONG2NUM(-FIX2LONG(nexp)))))`?

File: symbol.c [hunk 12]
```diff
@@ -316,31 +316,35 @@ rb_enc_symname_type(const char *name, long len, rb_encoding *enc, unsigned int a
...
-	  case '=':
-	    if (!(allowed_attrset & (1U << type))) return -1;
-	    type = ID_ATTRSET;
-	    ++m;
-	    break;
-	}
-	break;
+        goto id;
+    }
+    goto stophere;
```
Comment (nobu) [inline symbol.c:321]: Is foxtrot like this really better?

File: string.c [hunk 13]
```diff
@@ -3667,34 +3669,19 @@ rb_str_index_m(int argc, VALUE *argv, VALUE str)
...
-	    rb_raise(rb_eTypeError, "type mismatch: %s given",
-		     rb_obj_classname(sub));
-	}
-	sub = tmp;
-      }
-	/* fall through */
-      case T_STRING:
+    }
+    else {
+        StringValue(sub);
```
Comment (nobu) [inline string.c:3682]: This changes the error message, didn't it matter?

File: time.c [hunk 14]
```diff
@@ -3276,8 +3265,8 @@ find_time_t(struct tm *tptr, int utc_p, time_t *tp)
     status = 1;
 
     while (guess_lo + 1 < guess_hi) {
+      binsearch:
         if (status == 0) {
-          binsearch:
```
Comment (nobu) [inline time.c:3280]: Can't jump here when `status != 0`?

**Comment by nobu** [comment]
Not all compilers understand `NORETURN` attribute, so `UNREACHABLE_RETURN` may be better after global jump at the end of non-void functions.

File: enum.c (hunk 4 above)
Comment (shyouhei) [reply enum.c:3854]: I was aware of that and thought it is okay (0+something must be something).  But on a second thought it might be a problem because `v` can be something like a String?

File: error.c (hunk 5 above)
Comment (shyouhei) [reply error.c:925]: Yes.  This is an error handling path.  Should not bother runtime performance.

But might better change `builtin_class_name` to return fstr?

File: eval.c (hunk 6 above)
Comment (shyouhei) [reply eval.c:821]: :+1: Yes, that's a nice idea.

File: eval.c (hunk 7 above)
Comment (shyouhei) [reply eval.c:829]: `argc > 1` should also left `mesg` be `Qnil`.

**Comment by shyouhei** [comment]
Agreed on `NORETURN` comments.  Will add them.

File: object.c (hunk 10 above)
Comment (shyouhei) [reply object.c:3382]: Yes I think nobody does that.

File: rational.c (hunk 11 above)
Comment (shyouhei) [reply rational.c:2442]: OK, I take that.  I have no strong opinion.

File: symbol.c (hunk 12 above)
Comment (shyouhei) [reply symbol.c:321]: I agree this is something not very clean.

File: time.c (hunk 14 above)
Comment (shyouhei) [reply time.c:3280]: Maybe we need to add `status = 0;` just before `goto binsearch;`

File: enum.c (hunk 4 above)
Comment (shyouhei) [reply enum.c:3854]: Note: Strings cannot reach this path.

**Comment by headius** [comment]
FWIW I believe the implementation I did for JRuby linked above is fully compatible and uses no gotta, but it does still simulate them using a sort of state machine. This code has been used heavily in JRuby for many years and I don't think there are any outstanding issues.

File: string.c (hunk 13 above)
Comment (shyouhei) [reply string.c:3682]: At least there is no test who checks this error message.

Comment (nobu) [reply string.c:3683]: OK, just a confirmation.

File: compile.c (hunk 2 above)
Comment (shyouhei) [reply compile.c]: https://github.com/ruby/ruby/pull/3247/commits/d472f2652b2d39d359741d788fa4a23384f9aa32

File: error.c (hunk 5 above)
Comment (shyouhei) [reply error.c]: https://github.com/ruby/ruby/pull/3247/commits/59ae2d5232be64f118d9b01b131c7a63a5d716e0

File: eval.c (hunk 6 above)
Comment (shyouhei) [reply eval.c:821]: https://github.com/ruby/ruby/pull/3247/commits/c0bc23f6e0fe0628da8fd89d8ba31fbd89d339ff

File: dir.c (hunk 3 above)
Comment (shyouhei) [reply dir.c]: https://github.com/ruby/ruby/pull/3247/commits/12a56d74cad9ae2158a1485023ecee590ac67a73

File: symbol.c (hunk 12 above)
Comment (shyouhei) [reply symbol.c]: https://github.com/ruby/ruby/pull/3247/commits/12a56d74cad9ae2158a1485023ecee590ac67a73

File: time.c (hunk 14 above)
Comment (shyouhei) [reply time.c:3280]: https://github.com/ruby/ruby/pull/3247/commits/17f0f3b8803d14179dc8b44c79cecf077dee586e

**Comment by shyouhei** [comment]
@nobu I think converting a goto into an enum is not enough to be "better".  Returning that enum from a function is the key point.

**Comment by nobu** [comment]
Isn't it out of scope of this PR?

**Comment by shyouhei** [comment]
@nobu Kind of yes, but you concerned about the "better" perspective, no? https://github.com/ruby/ruby/pull/3247/commits/b62aeca1e1baf0f3acd26469df8645b7f8a07e89 should be enough then.

**Comment by nobu** [comment]
b62aeca would be enough.

First, I agree that `goto` into a branch is not a good idea in general.
But such `goto`s used currently are mostly for escape from functions and placed other than the end just to suppress some compiler warnings, so I don't think they are really problematic.

**Comment by shyouhei** [comment]
To save everybody's weekend I'll merge this on Monday.

**Comment by boazsegev** [comment]
**TL;DR;**: the effect of inlining error branches (instead of using `goto`) could **degrade performance**.

Moving error brunches to the end of a function (using `goto`) is a **similar hint to using `unlikely`** (used extensively in Linux kernel code) and could result in **faster code** due to possible cache locality.

Many developers ignore this use-case and leave such optimizations to the compiler (or ignore them). However, depending on the execution environment **this PR could have a negative effect on performance**.

In the most common case, performance effects should be too small to matter - however, some execution environments might suffer more severely (depending on cache line length and code memory addresses, [things that can occur on a per-execution basis](https://youtu.be/r-TLSBdHe1A?t=600)).

**Comment by shyouhei** [comment]
I bet the situation is more complicated than you think (or any other human beings think; CPUs are smarter than us nowadays), but yes the changes provided here could impact runtime performance.  I'm not sure if such the impact is positive or negative though.
